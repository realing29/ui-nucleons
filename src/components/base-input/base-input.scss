@import '../colors';

.reset {
  border: 0;
  box-sizing: border-box;
  margin: 0;
  outline: 0;
  padding: 0;
  font: inherit;
}

.root {
  position: relative;
  background-color: transparent;
  width: 0;
  flex-grow: 1;
  display: flex;
}

.field {
  width: 0;
  flex-grow: 1;
  display: block;
  color: inherit;
  font-size: var(--font-size);
  line-height: var(--line-height);
  z-index: 1;
  text-overflow: inherit;
  background: inherit;
  &::placeholder {
    opacity: 1;
    color: var(--placeholder-color);
  }
  &::-ms-clear { // отключаем системный крестик в IE
    display: none;
    height: 0;
    width: 0;
  }
  &.multiline {
    resize: none;

    // скрытие полосы прокрутки
    scrollbar-width: none; // Firefox
    overflow: -moz-scrollbars-none; // old Firefox
    -ms-overflow-style: none; // IE, old Edge
    &::-webkit-scrollbar {
      display: none; // Chrome, Safari, Opera
    }
  }
}

.fake-text {
  pointer-events: none;
  position: absolute;
  width: 100%;
  height: 100%;
  font-family: inherit;
  font-size: var(--font-size);
  line-height: var(--line-height);
  overflow: hidden;
  z-index: 0;
  .invisible-value {
    color: transparent;
  }
  .placeholder {
    color: var(--placeholder-color);
  }
}
