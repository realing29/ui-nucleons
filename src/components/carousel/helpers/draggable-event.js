/**
 * Нестандартное событие перетаскивания.
 */
export default class DraggableEvent {
  /**
   * Конструктор.
   * @param {Object} props Свойства.
   * @param {import('../../helpers/point').Point} props.offset Смещение контента.
   * @param {import('../../helpers/point').Point} props.client Клиентская позиция.
   */
  constructor (props) {
    this.prevented = false;
    this.offset = props.offset;
    this.client = props.client;
  }

  /**
   * Помечает событие как отмененное.
   */
  preventDefault () {
    this.prevented = true;
  }
}
