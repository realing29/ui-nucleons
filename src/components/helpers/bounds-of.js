/**
 * Возвращает ограничивающий прямоугольник элемента.
 * @param {Element} [element] Элемент.
 * @return {DOMRect|null} Ограничивающий прямоугольник.
 */
const boundsOf = element => element && element.getBoundingClientRect
  ? element.getBoundingClientRect()
  : null;

export default boundsOf;
