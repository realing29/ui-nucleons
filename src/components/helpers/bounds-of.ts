/**
 * Возвращает ограничивающий прямоугольник элемента.
 * @param element Элемент.
 * @return Ограничивающий прямоугольник.
 */
const boundsOf = (element?: HTMLElement) => element && element.getBoundingClientRect
  ? element.getBoundingClientRect()
  : null;

export default boundsOf;
