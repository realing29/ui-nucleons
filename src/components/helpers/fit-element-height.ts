/**
 * Перехватив событие, подгонит высоту поля под содержимое.
 * @param {Event} event Событие.
 * @param {HTMLElement} event.target Поле.
 */
export const fitElementHeight = ({ target }: { target: HTMLElement }) => {
  target.style.height = 'auto';

  // актуальные значения после установки "auto" (в начало функции не переносить)
  const {
    scrollHeight,
    offsetHeight,
    clientHeight,
  } = target;

  target.style.height = `${scrollHeight + offsetHeight - clientHeight}px`;
};
