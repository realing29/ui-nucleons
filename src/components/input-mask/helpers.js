import toArray from 'lodash/toArray';
import findLastIndex from 'lodash/findLastIndex';

/**
 * Возвращает начальное значение инпута.
 * @param {Object} params Переданные параметры.
 * @param {boolean} params.alwaysShowMask Нужно ли всегда показывать маску.
 * @param {string} params.value Значение инпута.
 * @param {string} params.mask Маска.
 * @param {RegExp} params.pattern Паттерн допустимых значений.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @return {string} Начальное значение инпута.
 */
export const getStartValue = ({ alwaysShowMask, value, mask, pattern, maskSign }) => {
  let startValue;

  if (value || alwaysShowMask) {
    startValue = mask
      ? prepareValue({ value, mask, maskSign, pattern })
      : (value.match(pattern) || []).join('');
  }

  return startValue || '';
};

/**
 * Преобразование значения под маску.
 * @param {Object} params Переданные параметры.
 * @param {string} params.value Значение инпута.
 * @param {string} params.mask Маска.
 * @param {RegExp} params.pattern Паттерн допустимых значений.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @return {string} Преобразованное значение.
 */
export const prepareValue = ({ value, mask, maskSign, pattern }) => {
  const patternMask = (mask.match(pattern) || []).join('');
  const patternValueArray = value.match(pattern) || [];
  const patternValue = patternValueArray.join('');

  // Убираем цифры маски из переданного значения инпута.
  const lengthPatternMask = patternMask.length;
  lengthPatternMask
    && patternValue.indexOf(patternMask) === 0
    && patternValueArray.splice(0, lengthPatternMask);

  // Последовательно заполняем заменяемые символы маски по возможности.
  const correctValue = mask
    ? toArray(mask).map(character => {
      let newCharacter;
      if (character === maskSign) {
        newCharacter = patternValueArray.shift();
      }
      return newCharacter || character;
    }).join('')
    : patternValue;

  return correctValue || mask;
};

/**
 * Возвращает корректную позицию курсора.
 * @param {Object} params Переданные параметры.
 * @param {string} params.value Значение инпута.
 * @param {string} params.mask Маска.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @param {number} params.position Переданное положение курсора.
 * @return {number} Позиция курсора.
 */
export const validateCursorPosition = ({ value, mask, maskSign, position }) => {
  const firstValueCaret = value.indexOf(maskSign);
  const firstMaskCaret = mask.indexOf(maskSign);
  let correctPosition = position >= 0 ? position : 0;

  if (firstValueCaret >= 0 && correctPosition >= firstValueCaret) {
    correctPosition = firstValueCaret;
  }

  if (firstMaskCaret >= 0 && correctPosition <= firstMaskCaret) {
    correctPosition = firstMaskCaret;
  }

  return correctPosition;
};

/**
 * Обрабатывает изменения инпута.
 * @param {Object} params Переданные параметры.
 * @param {HTMLElement} params.input Инпут.
 * @param {string} params.mask Маска.
 * @param {RegExp} params.pattern Паттерн допустимых значений.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @param {null|number} params.lastSelection Последнее значение положения курсора.
 * @return {Object} Обновленные данные значения инпута и положения курсора.
 */
export const processChange = ({ input, mask, maskSign, pattern, lastSelection }) => {
  const { value = '', selectionStart } = input || {};
  const maskArray = toArray(mask);
  const valueCharactersArray = toArray(value);
  let position = selectionStart;

  const isAdded = value.length > mask.length;

  // Расчет позиции курсора, чтоб он всегда находился перед заменяемым знаком маски.
  if (maskArray[selectionStart] !== maskSign) {
    const deltaSelection = lastSelection - selectionStart;
    const [needRemoveBefore, needRemoveAfter] = [deltaSelection === 1, deltaSelection === 0];

    if (needRemoveBefore) {
      position = findLastIndex(maskArray, (character, index) =>
        index < selectionStart && character === maskSign);
    } else if (needRemoveAfter || isAdded) {
      position = maskArray.findIndex((character, index) =>
        index >= selectionStart && character === maskSign);
    }

    // Удаление корректного значения, если при удалении попали на пробел/скобку и т.д.
    if (position >= 0 && (needRemoveBefore || needRemoveAfter)) {
      const removeIndex = needRemoveBefore ? position : position - 1;
      valueCharactersArray[removeIndex] = maskArray[removeIndex];
    }
  }

  const correctValue = prepareValue({
    value: valueCharactersArray.join(''),
    mask,
    maskSign,
    pattern,
  }) || '';

  return {
    correctValue,
    correctSelection: validateCursorPosition({
      value: correctValue,
      mask,
      maskSign,
      position: position > 0 ? position : correctValue.length,
    }),
  };
};
