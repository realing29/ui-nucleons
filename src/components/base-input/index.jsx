import React, { useRef, useEffect, useImperativeHandle, forwardRef } from 'react';
import classnames from 'classnames/bind';
import classes from './base-input.scss';
import PropTypes from 'prop-types';
import { fitElementHeight } from '../helpers/fit-element-height';
import isFunction from 'lodash/isFunction';
import { isString } from 'lodash';

const cx = classnames.bind(classes);

/**
 * Компонент поля ввода.
 * @param {Object} props Свойства. Поддерживаются свойства элемента input.
 * @param {boolean} [props.disabled] Отключено ли поле.
 * @param {string} [props.className] CSS-класс.
 * @param {Function} [props.onInput] Сработает при событии input.
 * @param {boolean} [props.multiline] Нужно ли выводить textarea вместо input.
 * @param {number|string} [props.rows] Значение атрибута rows для textarea.
 * @param {string|Object} props.style Стили.
 * @param {string} props.value Значение.
 * @return {ReactElement} Компонент поля ввода.
 */
export const BaseInput = forwardRef(function BaseInput ({
  className,
  disabled,
  multiline,
  onInput,
  restPlaceholder: restPlaceholderProp,
  rows = 1,
  style,
  value,
  ...props
}, ref) {
  const Element = multiline ? 'textarea' : 'input';
  const inputRef = useRef();

  useImperativeHandle(ref, () => inputRef.current);

  useEffect(() => {
    multiline && inputRef.current && fitElementHeight({
      target: inputRef.current,
    });
  }, []);

  const restPlaceholder = isString(restPlaceholderProp)
    ? { shiftValue: value, value: restPlaceholderProp }
    : { shiftValue: value, ...restPlaceholderProp };

  return (
    <div
      style={style}
      className={cx(
        'reset',
        'root',
        disabled && 'disabled',
        className,
      )}
    >
      {!multiline && Boolean(restPlaceholder.value) && (
        <span className={cx('fake-text')}>
          <span className={cx('invisible-value')}>{restPlaceholder.shiftValue}</span>
          <span className={cx('placeholder')}>{restPlaceholder.value}</span>
        </span>
      )}
      <Element
        {...props}
        rows={multiline ? rows : undefined}
        ref={inputRef}
        disabled={disabled}

        // ВАЖНО: не даем возможности задать стили/классы именно для этого элемента
        className={cx(
          'field',
          'reset',
          multiline && 'multiline',
        )}
        value={value}
        onInput={event => {
          multiline && fitElementHeight(event);
          isFunction(onInput) && onInput(event);
        }}
      />
    </div>
  );
});

BaseInput.propTypes = {
  /**
   * Отключено ли поле.
   */
  disabled: PropTypes.bool,

  /**
   * CSS-класс.
   */
  className: PropTypes.string,

  /**
   * Сработает при событии input.
   */
  onInput: PropTypes.func,

  /**
   * Нужно ли выводить textarea вместо input.
   */
  multiline: PropTypes.bool,

  /**
   * Значение атрибута rows для textarea.
   */
  rows: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

  /**
   * Placeholder, который выводится после введенного значения.
   */
  restPlaceholder: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({ shiftValue: PropTypes.string, value: PropTypes.string.isRequired }),
  ]),

  /**
   * Стили.
   */
  style: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  /**
   * Введенное значение.
   */
  value: PropTypes.string,
};
