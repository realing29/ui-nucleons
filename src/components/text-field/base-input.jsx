import React, { useRef, useEffect, useImperativeHandle, forwardRef } from 'react';
import { cx } from './classes';
import PropTypes from 'prop-types';
import { fitElementHeight } from '../helpers/fit-element-height';
import isFunction from 'lodash/isFunction';

/**
 * Компонент поля ввода.
 * @param {Object} props Свойства. Поддерживаются свойства элемента input.
 * @param {'xs'|'s'|'l'} props.size Размер.
 * @param {boolean} [props.failed] Ошибочно ли введено значение.
 * @param {boolean} [props.disabled] Отключено ли поле.
 * @param {string} [props.className] CSS-класс.
 * @param {Function} [props.onInput] Сработает при событии input.
 * @param {boolean} [props.multiline] Нужно ли выводить textarea вместо input.
 * @param {number|string} [props.rows] Значение атрибута rows для textarea.
 * @return {ReactElement} Компонент поля ввода.
 */
export const BaseInput = forwardRef(function BaseInput ({
  size,
  failed,
  disabled,
  className,
  onInput,
  multiline,
  rows = 1,
  ...props
}, ref) {
  const Element = multiline ? 'textarea' : 'input';
  const inputRef = useRef();

  useImperativeHandle(ref, () => inputRef.current);

  useEffect(() => {
    multiline && inputRef.current && fitElementHeight({
      target: inputRef.current,
    });
  }, []);

  return (
    <Element
      {...props}
      rows={multiline ? rows : undefined}
      ref={inputRef}
      disabled={disabled}
      className={cx([
        className,
        'reset',
        'base-input',
        failed && 'failed',
        disabled && 'disabled',
        multiline && 'multiline',
        size && `size-${size}`,
      ])}
      onInput={event => {
        multiline && fitElementHeight(event);
        isFunction(onInput) && onInput(event);
      }}
    />
  );
});

BaseInput.propTypes = {
  /**
   * Размер.
   */
  size: PropTypes.oneOf(['xs', 's', 'l']),

  /**
   * Ошибочно ли введено значение.
   */
  failed: PropTypes.bool,

  /**
   * Отключено ли поле.
   */
  disabled: PropTypes.bool,

  /**
   * CSS-класс.
   */
  className: PropTypes.string,

  /**
   * Сработает при событии input.
   */
  onInput: PropTypes.func,

  /**
   * Нужно ли выводить textarea вместо input.
   */
  multiline: PropTypes.bool,

  /**
   * Значение атрибута rows для textarea.
   */
  rows: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};
