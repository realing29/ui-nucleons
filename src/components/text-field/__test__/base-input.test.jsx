import React, { createRef } from 'react';
import { mount } from 'enzyme';
import { BaseInput } from '../base-input';
import { fitElementHeight } from '../../helpers/fit-element-height';

jest.mock('../../helpers/fit-element-height', () => {
  const original = jest.requireActual('../../helpers/fit-element-height');

  return {
    ...original,
    __esModule: true,
    fitElementHeight: jest.fn(original.fitElementHeight),
  };
});

describe('<BaseInput />', () => {
  it('should render without props', () => {
    const wrapper = mount(
      <BaseInput />
    );

    expect(wrapper).toMatchSnapshot();
  });
  it('should forward ref', () => {
    const testRef = createRef();

    const wrapper = mount(
      <BaseInput ref={testRef} />
    );

    expect(wrapper.getDOMNode().nodeName).toBe('INPUT');
    expect(testRef.current).toBe(wrapper.getDOMNode());
  });
  it('should call "fitElementHeight" after mount', () => {
    expect(fitElementHeight).toHaveBeenCalledTimes(0);
    mount(<BaseInput multiline />);
    expect(fitElementHeight).toHaveBeenCalledTimes(1);
  });
});
