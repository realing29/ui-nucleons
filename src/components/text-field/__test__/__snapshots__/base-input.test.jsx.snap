// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseInput /> should render without props 1`] = `
<ForwardRef(BaseInput)>
  <input
    className="reset base-input"
    onInput={[Function]}
  />
</ForwardRef(BaseInput)>
`;
